### ISSUE TITLE
Good Range for learning rate?

### ISSUE THREAD

**Issue by PaulForInvent** [UKPLab/sentence-transformers#791]
Hey, what is a good range to test different learning rates and epsilons?  Is it advisible to focus on the scale e-5 to e-6? Would you tune on a linear or log scale?

Thank you!

**Comment by nreimers** [comment]
Have a look at the BERT paper, there they recommend 3 learning rates that work quite well. Learning rate does not have the largest impact and can be kept as is. 

Never experimented with epsilon, but I think the impact will also not be significant

**Comment by PaulForInvent** [comment]
Thanks. 

> Learning rate does not have the largest impact and can be kept as is.

What else will be more important, specifically from the optimizer? I think of epoch, batch size, but thought of some optimizer parameters too?

**Comment by nreimers** [comment]
These usually have rather small impacts.

What is most important:
- Quality of your training data
- The type of loss your are using (and associated hyperparameters like batch size)
- Having good hard negatives when using TripletLoss or MultipleNegativeRankingLoss etc
- The underlying transformer model (BERT vs RoBERTa etc.)

**Comment by PaulForInvent** [comment]
@nreimers  I can totally agree with these.

> * Having good hard negatives when using TripletLoss or MultipleNegativeRankingLoss etc

So for the case of MultipleNegativeRankingLoss , you have to consider to 1) exclude same positve anchors inside the btch and 2 ) and also make sure having hard negatives within this batch? So 2 things you have to customize for your loss function.

**Comment by nreimers** [comment]
For MultipleNegativeRankingLoss , you can provide input examples in the format:
(anchor, positive, hard_negative_1, hard_negative_2, ...)

All examples in a batch must have the same number of hard negatives.

**Comment by tide90** [comment]
@nreimers 

> For MultipleNegativeRankingLoss , you can provide input examples in the format:
> (anchor, positive, hard_negative_1, hard_negative_2, ...)
> 
> All examples in a batch must have the same number of hard negatives.

It is not really clear how to use it with hard negatives.

For multiple hard negatives, there can be 2 ways, right?

1.) yours:  (anchor, positive, hard_negative_1, hard_negative_2, ...)
2) or seperately:  
(anchor, positive, hard_negative_1)
 (anchor, positive, hard_negative_2)
 (anchor, positive, hard_negative_3)
...

And, what if I have different number of hard negatives for each sample?

I suppose only for option1 the number per batch need to be the same, so I just use option 2 for various numbers? Somehow confusing.

I hope you could make it more clear! 👍

**Comment by nreimers** [comment]
Each example in a batch must have the same number of hard negatives.

**Comment by tide90** [comment]
> Each example in a batch must have the same number of hard negatives.

But this does not answer I I can use hard negative to include them... As I asked, maybe you could clarify above 2 options.

I want to use this loss with hard negative for days now, but really do not know how to do this...

@nreimers

**Comment by nreimers** [comment]
Yes, use option 2. That is the better option.

**Comment by tide90** [comment]
Do you mean each example at all independent of their class? Or per class need to be same number?

**Comment by tide90** [comment]
> Do you mean each example at all independent of their class? Or per class need to be same number?

@nreimers : And for multiple hard negatives: If I pass them via option 2 I can have different numbers, right? SO only if I pass them via option 1 the numbers have to be same per batch?

**Comment by nreimers** [comment]
Correct

**Comment by tide90** [comment]
> Correct

@nreimers 

It seems not correct. I just tried it. Also if you put examples individually, you cannot have differnet numbers. Also, if you want to use hard negatives ALL training examples have to be with a negative example.

So you cannot leave some example without a negative or having multiple negatives for only some examples...

So, also via option 2, you cannot have different numbers of negatives. See last comment.

Sorry, but could you write up the correct usage inside the loss function in the doc string?

**Comment by dreamflasher** [comment]
From https://wandb.ai/jack-morris/david-vs-goliath/reports/Does-Model-Size-Matter-A-Comparison-of-BERT-and-DistilBERT--VmlldzoxMDUxNzU:
```
The BERT authors [recommend](https://github.com/google-research/bert) fine-tuning for 4 epochs over the following hyperparameter options:
batch sizes: 8, 16, 32, 64, 128
learning rates: 3e-4, 1e-4, 5e-5, 3e-5
```

```
1. Relative to batch size, learning rate has a much higher impact on model performance. So if you're choosing to search over potential learning rates and potential batch sizes, it's probably wiser to search spend more time tuning the learning rate.
2. The learning rate has a very high negative correlation (-0.540) with model accuracy. Therefore, in our search space, lower learning rates are better.
```

```
Our data indicate that the best hyperparameters for fine-tuning BERT are batch_size 64 and learning rate 5e-5
```


From the BERT paper (https://arxiv.org/pdf/1810.04805.pdf):
```
The optimal hyperparameter values are task-specific, but we found the following range of possible values to work well across all tasks:
* Batch size: 16, 32
* Learning rate (Adam): 5e-5, 3e-5, 2e-5
* Number of epochs: 2, 3, 4
We also observed that large data sets (e.g., 100k+ labeled training examples) were far less sensitive to hyperparameter choice than small data sets.
```
